openlibrary/plugins/openlibrary/js/carousels.js

```diff
@@ -82,40 +80,77 @@ const Carousel = {
+            } catch (e) {
+                url = new URL(window.location.origin + loadMore.url);
+            }
+            var default_limit = 18; // 3 pages of 6 books
+            url.searchParams.set('limit', loadMore.limit || default_limit);
+            loadMore.pageMode = loadMore.pageMode === 'page' ? 'page' : 'offset'; // verify pagination mode
+            loadMore.locked = false; // prevent additional calls when not in critical section
+
+            // Bind an action listener to this carousel on resize or advance
+            $(selector).on('afterChange', function() {
+                var totalSlides = $(selector + '.slick-slider')
+                    .slick("getSlick").$slides.length;
+                var numActiveSlides = $(selector + ' .slick-active').length;
+                var currentLastSlide = $(selector + '.slick-slider')
+                    .slick('slickCurrentSlide') + numActiveSlides;
+                // this allows us to pre-load before hitting last page
+                var lastSlideOn2ndLastPage = (totalSlides - numActiveSlides);
 
-            if (loadMoreUrl && (currentLastSlide >= lastSlideOnSecondToLastPage)) {
-                var limit = numActiveSlides * 3;
-                var url = loadMoreUrl + '?offset=' + totalSlides + '&limit=' + limit;
-                $.ajax({
-                    'url': url,
-                    'type': 'GET',
-                    success: function(subject_results) {
-                        $.each(subject_results.works, function(work_idx) {
-                            var work = subject_results.works[work_idx];
-                            var lastSlidePos = $('.carousel-' + key + '.slick-slider').slick("getSlick").$slides.length - 1;
-                            $('.carousel-' + key).slick('slickAdd', addWork(work), lastSlidePos);
-                        });
+                if (!loadMore.locked && (currentLastSlide >= lastSlideOn2ndLastPage)) {
+                    loadMore.locked = true; // lock for critical section
+                    document.body.style.cursor='wait'; // change mouse to spin
+
+                    if (loadMore.pageMode == 'page') {
+                        // for first time, we're on page 1 already so initialize as page 2
+                        // otherwise advance to next page
+                        loadMore.page = loadMore.page ? loadMore.page + 1 : 2;
+                    } else { // i.e. offset, start from last slide
+                        loadMore.page = totalSlides;
                     }
-                });
-            }
-        });
+
+                    // update the current page or offset within the URL
+                    url.searchParams.set(loadMore.pageMode, loadMore.page);
+
+                    $.ajax({
+                        'url': url,
+                        'type': 'GET',
+                        success: function(subject_results) {
+                            $.each(subject_results.works, function(work_idx) {
```

Comment:
Why not use Array.prototype.forEach here?
The jQuery each method adds a lot of indirection.

Reply:
This also should also be broken out into a separate function and tested. I have no idea what this is doing and am concerned that the data being returned from the API could break if anything mandatory is missing resulting in the failure to render any carousel items.